pkg/proxy/conntrack/cleanup.go

```diff
@@ -59,57 +62,124 @@ func deleteStaleServiceConntrackEntries(ct Interface, svcPortMap proxy.ServicePo
 			}
 			for _, lbIP := range svcInfo.LoadBalancerVIPs() {
-				err := ct.ClearEntriesForNAT(lbIP.String(), endpointIP, v1.ProtocolUDP)
-				if err != nil {
-					klog.ErrorS(err, "Failed to delete endpoint connections for LoadBalancerIP", "servicePortName", epSvcPair.ServicePortName, "loadBalancerIP", lbIP)
-				}
+				filters = append(filters, filterForNAT(lbIP.String(), endpointIP, v1.ProtocolUDP))
 			}
 		}
 	}
+
+	if err := ct.ClearEntries(getUnixIPFamily(ipFamily), filters...); err != nil {
+		klog.ErrorS(err, "Failed to delete stale endpoint connections")
+	}
+}
+
+// getUnixIPFamily returns the unix IPFamily constant.
+func getUnixIPFamily(ipFamily v1.IPFamily) uint8 {
+	if ipFamily == v1.IPv6Protocol {
+		return unix.AF_INET6
+	}
+	return unix.AF_INET
+}
+
+// getProtocolNumber return the Assigned Internet Protocol Number.
+// https://www.iana.org/assignments/protocol-numbers/protocol-numbers.xhtml
+func getProtocolNumber(proto v1.Protocol) uint8 {
+	switch proto {
+	case v1.ProtocolTCP:
+		return unix.IPPROTO_TCP
+	case v1.ProtocolUDP:
+		return unix.IPPROTO_UDP
+	case v1.ProtocolSCTP:
+		return unix.IPPROTO_SCTP
+	}
+	return 0
+}
+
+// filterForIP returns *conntrackFilter to delete the conntrack entries for connections
+// specified by the destination IP(original direction).
+func filterForIP(ip string, protocol v1.Protocol) *conntrackFilter {
+	klog.V(4).InfoS("Adding conntrack filter for cleanup", "org-dst", ip, "protocol", protocol)
+	return &conntrackFilter{
+		protocol: getProtocolNumber(protocol),
+		original: &connectionTuple{
+			dstIP: netutils.ParseIPSloppy(ip),
+		},
+	}
+}
+
+// filterForPort returns *conntrackFilter to delete the conntrack entries for connections
+// specified by the destination Port(original direction).
+func filterForPort(port int, protocol v1.Protocol) *conntrackFilter {
+	klog.V(4).InfoS("Adding conntrack filter for cleanup", "org-port-dst", port, "protocol", protocol)
+	return &conntrackFilter{
+		protocol: getProtocolNumber(protocol),
+		original: &connectionTuple{
+			dstPort: uint16(port),
+		},
+	}
+}
+
+// filterForNAT returns *conntrackFilter to delete the conntrack entries for connections
+// specified by the destination IP(original direction) and source IP(reply direction).
+func filterForNAT(origin, dest string, protocol v1.Protocol) *conntrackFilter {
+	klog.V(4).InfoS("Adding conntrack filter for cleanup", "org-dst", origin, "reply-src", dest, "protocol", protocol)
+	return &conntrackFilter{
+		protocol: getProtocolNumber(protocol),
+		original: &connectionTuple{
+			dstIP: netutils.ParseIPSloppy(origin),
+		},
+		reply: &connectionTuple{
+			srcIP: netutils.ParseIPSloppy(dest),
+		},
+	}
+}
+
+// filterForPortNAT returns *conntrackFilter to delete the conntrack entries for connections
+// specified by the destination Port(original direction) and source IP(reply direction).
+func filterForPortNAT(dest string, port int, protocol v1.Protocol) *conntrackFilter {
```

Comment:
note to self, this is replacement for

```
func (ct *execCT) ClearEntriesForPortNAT(dest string, port int, protocol v1.Protocol) error {
	if port <= 0 {
		return fmt.Errorf("wrong port number. The port number must be greater than zero")
	}
	parameters := parametersWithFamily(utilnet.IsIPv6String(dest), "-D", "-p", protoStr(protocol), "--dport", strconv.Itoa(port), "--dst-nat", dest)
```

Reply:
@aroradaman this is confusing me, original is `--dst-nat`  that in the library filter is

https://github.com/aroradaman/netlink/blob/c96b03b4befaf5cc418386ad9eaf1a5d15101b1e/conntrack_linux.go#L842-L844

```
	// -dst-nat ip    Destination NAT ip
		if elem, found := f.ipNetFilter[ConntrackReplyDstIP]; match && found {
			match = match && elem.Contains(flow.Reverse.DstIP)
		}
```

are we sure are we not switching the directions here? should it be reply-dst?

you also have it this way in the filter https://github.com/kubernetes/kubernetes/pull/126847/files#r1742000074

Reply:
heh, confusing

```
root@kind-worker:/# conntrack -L --reply-dst 192.168.8.5
conntrack v1.4.7 (conntrack-tools): 0 flow entries have been shown.
root@kind-worker:/# conntrack -L --reply-src 192.168.8.5
tcp      6 86392 ESTABLISHED src=192.168.8.7 dst=192.168.8.5 sport=43656 dport=6443 src=192.168.8.5 dst=192.168.8.7 sport=6443 dport=43656 [ASSURED] mark=0 use=1
tcp      6 86392 ESTABLISHED src=192.168.8.7 dst=192.168.8.5 sport=43618 dport=6443 src=192.168.8.5 dst=192.168.8.7 sport=6443 dport=43618 [ASSURED] mark=0 use=1
tcp      6 86389 ESTABLISHED src=10.244.1.3 dst=10.96.0.1 sport=58574 dport=443 src=192.168.8.5 dst=192.168.8.7 sport=6443 dport=58574 [ASSURED] mark=0 use=1
tcp      6 86392 ESTABLISHED src=192.168.8.7 dst=192.168.8.5 sport=43594 dport=6443 src=192.168.8.5 dst=192.168.8.7 sport=6443 dport=43594 [ASSURED] mark=0 use=1
conntrack v1.4.7 (conntrack-tools): 4 flow entries have been shown.
root@kind-worker:/# conntrack -L --dst-nat 192.168.8.5
tcp      6 86398 ESTABLISHED src=10.244.1.3 dst=10.96.0.1 sport=58574 dport=443 src=192.168.8.5 dst=192.168.8.7 sport=6443 dport=58574 [ASSURED] mark=0 use=1
conntrack v1.4.7 (conntrack-tools): 1 flow entries have been shown.
root@kind-worker:/# conntrack -L --src-nat 192.168.8.5
conntrack v1.4.7 (conntrack-tools): 0 flow entries have been shown.
```

Reply:
it seems conntrack userspace tool is able to differentiate between natted connection and normal connection

Reply:
I modelled the filters based on `netlink.ConntrackFlow` that doesn't have --src-nat and --dst-nat options.
For proxying, DNAT, --reply-src will always be the endpoint.

https://github.com/aroradaman/netlink/blob/c96b03b4befaf5cc418386ad9eaf1a5d15101b1e/conntrack_linux.go#L266

Reply:
ok, it was also reported in https://github.com/vishvananda/netlink/pull/374 🤦 

Do we need to dump only the NAT ones? I think this is fine, just we need to be completely sure we are deleting the right flow that I think is correct, teh reply source should be the endpoint IP

Reply:
validated